### ISSUE TITLE
Increase speed and reduce memory usage

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/imSim#145]
@adrianpope has access to some high level profiling codes at ANL that were already used to successfully profile and increase the throughput of PhoSim.

He can try to use these tools to identify hotspots in the code and let the imSim and GalSim teams (especially @rmjarvis) aware of the results.

To start writing down a few reference points: 

on a Centos VM with Intel(R) Xeon(R) CPU E5-2699 v4 @ 2.20GHz

I see running:

```
time imsim.py --psf Atmospheric --sensor 'R:2,2 S:1,1' --create_centroid_file --file_id test imsim_cat_197356.txt.gz
```

Where imsim_cat_197356.txt.gz is a reduced instance catalog with sources only on a few sensors:

```
real    88m17.320s
user    88m16.128s
sys    2m31.420s
```

Subdividing the run conditions adding more complexity each time I see:

```
15 min “bare”,
37 min + Atmosphere 
88 min + sensor
```

"bare" means you use --disable-sensor-model and -psf DoubleGaussian.
"+ Atmosphere"  means you change to the Atmospheric model.
"+ Sensor" means you enable the sensor model.

On KNL we see numbers 6-10 times this.  

It should also be noted that in order to build meas_base the MKL versions of the python library are disabled by default in the stack.

**Comment by rmjarvis** [comment]
I do think it's likely that using icpc to install GalSim will help at least somewhat if we're not doing that yet.  This in addition to trying to get the MKL-enabled numpy, which looks like a much bigger difference (according to ECP) than I would have expected.  I'd probably start with those two system-level changes before going into extreme detail about where the exact hot spots are.

**Comment by cwwalter** [comment]
Recording further previous discussions about memory usage:

Moving from GalSim 1.6 to GalSim 2.0 would reduce the memory used by atmospheric screens from 6 Gb to about 3 Gb. This memory is shared by all sensors running in a multiprocessor pool (multi-sensor mode in imSim).  According to @jmeyers314 possible further reduction in the memory used by the screens would be possible by (for example) moving from 64 to 32 bit floats.

**Comment by rmjarvis** [comment]
Note that the latter suggestion of moving to 32 bit floats would require some investigation that this doesn't overly degrade the accuracy of any calculations.  It's possible that there are numerical inaccuracies in some of our calculations that are not a problem when using double, but would become problematic if we switched to single precision.  

There may not be any problem, but I'm just pointing out that we can't blithely switch this data type without at least some investigation of the accuracy of the results.

**Comment by katrinheitmann** [comment]
Is it easy to estimate the amount of memory saved vs the amount of work 
needed for testing? If it's a lot of saving for a small testing efforts, 
it's definitely worth it ;-)

**Comment by cwwalter** [comment]
> Is it easy to estimate the amount of memory saved vs the amount of work
> needed for testing? If it's a lot of saving for a small testing efforts,
> it's definitely worth it ;-)

It would go from 3.0 to 1.5Gb. I can't estimate the work or testing :)

BTW:  for memory, the  matching number to my timing test above were:

```
1.7 Gb  “bare”
7.8 Gb + Atmosphere
8.1 Gb + Sensor
```

But these are less good numbers than the times since I was just trying to watch 'top'. Also note I was using a instance catalog that didn't have the whole focal planes worth of sources in it, just a few sensors.

**Comment by rmjarvis** [comment]
>  If it's a lot of saving for a small testing efforts, it's definitely worth it ;-)

Well, I tried it out and ran the GalSim unit tests, and many of the tests of the gradient in particular fail after the switch to single precision.  Some of these needed to back off to 1.e-3 relative accuracy in the tests, which is higher than we normally like.  (Especially since some of these have annotations along the lines of "Try a simpler interpolation function.  Derivatives should be exact.")  

So I think this means we'll need to do a more careful test that the atmospheric modules use of these gradients isn't going to have problems due to these numerical inaccuracies.

In other words, a "small" test has told me that we need a much longer test to see if this would be acceptable.  And I'd want help from @jmeyers314 to craft what the appropriate checks need to be for this.

**Comment by jmeyers314** [comment]
>Well, I tried it out and ran the GalSim unit tests, and many of the tests of the gradient in particular fail after the switch to single precision. 

That's interesting.  I'll take a look at the gradient tests with an eye for how relevant the unit tests may be or not be for atmospheric simulation in particular...

**Comment by jmeyers314** [comment]
I think it's a little hard to generalize from the 1.e-3 relative accuracy @rmjarvis mentioned, since presumably that's also a function of how finely gridded the table is.

Here's some quick math for atmospheric screen gradient requirements:  The deflection of a photon in radians is just gradient(wavefront), where the wavefront has dimensions of length and the gradient is therefore dimensionless.  For an accuracy of 1 milliarcsecond, we then need a gradient accuracy of 1e-3/206265 ~ 4e-9 (nm per nm).  This isn't actually that bad though considering the screens vary only by about +/- ~5000 nm over a distance of ~ 100 m, or 5000 nm / 1e11 nm ~ 5e-8.  So I think that means 10% relative accuracy may be good enough.  I think this is worth looking into further.

**Comment by rmjarvis** [comment]
I think mostly we need to devise an atmospheric PSF-specific test that the checks somehow that its use of LookupTable2D given the grid spacing it uses is acceptable.  To be fair, none of the existing PSF tests (in test_phase_screens.py) failed, so it might be already fine.  I just don't know that our unit tests suitable cover everything that might be affected by this change.

**Comment by jchiang87** [comment]
@rmjarvis 
Here are cProfile outputs for imSim at 81f3dcc5 and sim_GaSimInterface at [63b58f4](https://github.com/lsst/sims_GalSimInterface/commit/63b58f4fae99b07e4da93b4b006fcf14a6bc8e05) that I described at the CI meeting today:
* Two hour execution on cori-haswell, ~76k objects drawn:
[haswell_cprofile_2hours.txt](https://github.com/LSSTDESC/imSim/files/2458197/haswell_cprofile_2hours.txt)
* Execution through sky background rendering:
[haswell_cprofile_sky_bg.txt](https://github.com/LSSTDESC/imSim/files/2457579/haswell_cprofile_sky_bg.txt)
* Execution through Atmospheric PSF creation:
[haswell_cprofile_psf.txt](https://github.com/LSSTDESC/imSim/files/2457582/haswell_cprofile_psf.txt)

**Comment by jchiang87** [comment]
Sorry, I should have renamed them.  You can look at the contents like this (in python 3.6):
```
import pstats
stats = pstats.Stats('haswell_cprofile_2hours.txt')
stats.sort_stats('time').print_stats(20)
```

**Comment by rmjarvis** [comment]
Or, I just discovered this, which is pretty handy.  :)

https://www.stefaanlippens.net/python_profiling_with_pstats_interactive_mode/

**Comment by rmjarvis** [comment]
Here's one thing that's a bit funny.  It looks like the galsim `drawImage` is being called about 2x as often as `drawObject`:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2312/1    0.107    0.000 7198.810 7198.810 {built-in method builtins.exec}
        1    0.000    0.000 7198.809 7198.809 run_imsim.py:2(<module>)
        1    0.000    0.000 7004.542 7004.542 ImageSimulator.py:221(run)
        1    6.323    6.323 7004.542 7004.542 ImageSimulator.py:331(__call__)
    75732   18.852    0.000 6706.659    0.089 galSimInterpreter.py:788(drawObject)
   146213    7.409    0.000 4228.255    0.029 gsobject.py:1084(drawImage)
   146213    5.340    0.000 3848.926    0.026 gsobject.py:1899(drawPhot)
```
I assumed for DC2, we were usually running this with one bandpass and one detector at a time (in `self.bandpassDict` and `detectorList`), so I thought those for loops should have only been running through once.  

Is that not the case?  Are we somehow running multiple bands and/or detectors at once here?

Or do you see some other reason we would be calling drawImage twice for each object?

**Comment by rmandelb** [comment]
@rmjarvis - I was curious about this so I took a look at https://github.com/lsst/sims_GalSimInterface/blob/master/python/lsst/sims/GalSimInterface/galSimInterpreter.py
This could be naive, but it looks like drawObject first calls a routine findAllDetectors() to find what detectors a given object will hit, and that routine calls drawImage (but seemingly cheaply) before returning info back to drawObject and proceeding with the actual drawing.  This is the case even if you only pass in a single detector when instantiating the GalSimInterpreter(); it'll still check whether the object produces any light in that detector or not.

Sorry if this is just noise from the peanut gallery.  My working knowledge of this interface may be too sketchy to be useful, but this is what I got from reading the code.

**Comment by jchiang87** [comment]
@rmandelb I think that's the right explanation.  @danielsf could probably clarify further.

**Comment by rmandelb** [comment]
I will wait to see what @danielsf says, but my gut reaction (curious to hear Mike's thoughts) is that there must be a way to more optimally figure this out than drawing 1000 photons, perhaps using some of GalSim's routines for setting up images.

Note: since it's not the case that drawImage is called exactly twice the number of times drawObject() is called (it's a bit less than a factor of two), I presume that sometimes after running findAllDetectors() we're deciding we don't have to draw this object later on in drawObject().  So we can't just say "if there's only one detector passed in, don't use findAllDetectors" but we could optimize how we call drawImage here in findAllDetectors.  Unless this 2x calling is actually more costly than occasionally drawing ~10% more objects than needed?  (Not sure how that ~10% might fluctuate depending on the circumstances though.)

**Comment by rmjarvis** [comment]
Ah.  I didn't notice that call of drawImage.  Thanks.  

So, yeah, that's definitely not the most efficient way to do this.  Especially for the faint things, shooting 1000 photons is a big waste.  Even worse, you only use the size of the returned image, which means you didn't need to shoot any photons at all.
```
        centeredImage = centeredObj.drawImage(scale=testScale, method='phot', n_photons=1000, rng=self._rng)
        xmax = testScale*(centeredImage.xmax/2) + gsObject.xPupilArcsec
        xmin = testScale*(-1*centeredImage.xmax/2) + gsObject.xPupilArcsec
        ymax = testScale*(centeredImage.ymax/2) + gsObject.yPupilArcsec
        ymin = testScale*(-1*centeredImage.ymin/2) + gsObject.yPupilArcsec
```
All we actually use here is the image size, which is the same in x and y, and doesn't require either actually building the image nor shooting any photons.  This should be essentially equivalent:
```
        sizeArcsec = centeredObj.getGoodImageSize(1.0)  # pixel_scale = 1.0 means size is in arcsec.
        xmax = gsObject.xPupilArcsec + sizeArcsec/2.
        xmin = gsObject.xPupilArcsec - sizeArcsec/2.
        ymax = gsObject.yPupilArcsec + sizeArcsec/2.
        ymin = gsObject.yPupilArcsec - sizeArcsec/2.
```
There is some more code below that, which would also have to change, but I think everywhere you use `centeredImage`, it is only to use the either `xmax` or `ymax * testScale`.  So just use `sizeArcsec` instead.

**Comment by rmandelb** [comment]
@rmjarvis - I initially had the same thought, but then I noticed https://github.com/lsst/sims_GalSimInterface/blob/master/python/lsst/sims/GalSimInterface/galSimInterpreter.py#L234 and the line after, which do seem to use the flux of the drawn image?

**Comment by rmjarvis** [comment]
Well, I was wrong about the `folding_threshold` bit.  That's not done until later.  So we might be missing the bright objects that are off the edge.  The current code exacerbates this compared to my proposed code, but we might want to dramatically rethink how this is done to make sure we keep the bright things that will potentially overlap the detector.

Probably need to have different functions here for bright and faint things to make sure we are careful with bright things and don't waste a lot of time with faint things.

**Comment by rmjarvis** [comment]
BTW, for numbers on the time currently being spent in `findAllDetectors`:
```
    75732    7.683    0.000 1241.137    0.016 galSimInterpreter.py:161(findAllDetectors)
```
So 17% of the total time is spent in this function.  That's a lot more than should be required.

**Comment by rmjarvis** [comment]
https://github.com/lsst/sims_GalSimInterface/pull/68

**Comment by cwwalter** [comment]
It will also eventually be quite interesting to see if the same routines that take up most of time under KNL do the same on Haswell.  Hopefully, this work will also get us to the point where we can reasonably do an entire KNL job at some point.

**Comment by cwwalter** [comment]
I think that fractionally the sky calculation is not that large. You can see some numbers for the current code in Jim's slide 4 today from the CI meeting.

On Haswell I think ~< 10 minutes.  But x7.5 that on KNL which definitely isn't good.   But of course, that's still pretty small compared to the total run time.

**Comment by rmjarvis** [comment]
That was my guess too, but we also do flats and such, which are just sky calculations.  So I wasn't sure how much he cared about trying to improve this.

**Comment by cwwalter** [comment]
Good point on flats. 

For the coordutils/WCS if you did a high intensity flat that took much longer I wonder how much it would help much since you do those parts once right?

**Comment by jchiang87** [comment]
It's not the most important thing to address, but getting it to be as efficient as the galsim WCS would be nice.  For the flats, I actually use a bunch of [`GSFitsWCS` objects derived from template FITS files](https://github.com/LSSTDESC/desc_sim_utils/blob/master/bin.src/make_flats_multi.py#L52), which seemed a bit kludgey but necessary to make the (non-BF) flats in a reasonable amount of time.

**Comment by rmjarvis** [comment]
Here is a different item we could consider changing to speed up the imsim runs.
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    71209 1116.152    0.016 1116.152    0.016 sensor.py:247(accumulate)
```
This is the top item sorted by tottime, the code to accumulate the shot photons onto the sensor.

Internally, this code first has to read the existing flux on each pixel in the postage stamp, update all the pixel corners to account for the initial flux levels, and then add the shot photons to the right pixels.  This means that for objects with very few photons, it's dominated by that initial setup.

My proposal is for very faint things, just use a SimpleSensor, rather than the SiliconSensor, which doesn't do any B/F effect or tree rings, and so would be much faster.  I'm quite sure we don't care about the sensor effects for things with only a handful of photons (say <100).  Since these now make up a large fraction of the objects, this would help a lot.

The easiest way to do this is to just set `sensor=None` in the `drawImage` call if the flux is below a critical value rather than using `sensor=self.sensor[self.detector.name]`.

**Comment by cwwalter** [comment]
Hmm...well if an object had 100 photons in each exposure and was visited 800 times and was sitting close to a  bright object I think we could miss effects.

This is the sort of thing where it is easier to understand and convince yourself things are correct if you do everything the right way (similar to the argument Anze made about why he wanted to apply the PSF to the carpet of background objects).

I wonder if there could be a middle ground where we don't update any corners but at least put it into the already distorted pixels?

**Comment by rmjarvis** [comment]
That's the thing.  We don't have any distorted pixels until it runs that initial step to figure out how distorted the pixels are from the current image.

We could be slightly more careful by only skipping the silicon if the object has < 100 photons AND there is not yet any pixel in its postage stamp with > 100 e-.  We know the B/F effect doesn't start to matter until around 1000 e-/pixel, so these are both an order of magnitude conservative.

**Comment by rmjarvis** [comment]
Another idea that would require more refactoring is to try to group the faint objects in a patch of the sky, shoot all their photons into a single photon array.  Then accumulate them all at once.  

This would require significantly more code change in the sims code, and it would mean not just using the basic `drawImage` function.  We'd have to unpack that a bit to call `shoot` manually and build up a collection of photons from multiple objects.  But the tools to do that are all available in the GalSim API, so it's a possibility.

**Comment by rmandelb** [comment]
Sorry if I am missing something - in the above discussion I understand the relevance of being nearby to a really bright object, but what is the relevance of it being visited 800 times?  The b/f is driven by the number of photons hitting the sensor in a single visit, not the accumulation across all visits.  So I would think Mike's idea of "only skipping the silicon if the object has < 100 photons AND there is not yet any pixel in its postage stamp with > 100 e-" should be fairly conservative.  If we can tell the code to purposely do the faint objects in the sensor first, then that would be more of a win (since you wouldn't have a case where there's a lot of bright stuff that forces you to photon-shoot some really faint object that gets rendered later).

Mike, regarding this:
> Another idea that would require more refactoring is to try to group the faint objects in a patch of the 
> sky, shoot all their photons into a single photon array. Then accumulate them all at once.

Would that basically be a way to generate a faint galaxy carpet (going back to that idea) for a single sensor?

**Comment by cwwalter** [comment]
> orry if I am missing something - in the above discussion I understand the relevance of being nearby to a really bright object, but what is the relevance of it being visited 800 times? The b/f is driven by the number of photons hitting the sensor in a single visit, not the accumulation across all visits. So I would think Mike's idea of "only skipping the silicon if the object has < 100 photons AND there is not yet any pixel in its postage stamp with > 100 e-" should be fairly conservative.

Sorry.. I wasn't clear enough. I didn't mean that it added up (as you correctly point out).  What I meant was that you might only have a relatively few photons per visit so you might not think you could really notice this pretty small effect if you were looking at something with 100 photons in it.  But after after you do that sampling 800 times if there was a BF effect you would be able to see it (if it was happening at that level).

Also, I think thinking about this with a sky background of ~800 is a bit subtle.  It is true that the std dev of the number is small when you have a flat.  But as you build up the flat the BF effect is enough to distort the poisson statistics in this case since some pixels will be higher than others due to fluctuations.  So you also want to think about what you are doing on top of that background too.

**Comment by cwwalter** [comment]
Sorry hit send to early: so in "only skipping the silicon if the object has < 100 photons AND there is not yet any pixel in its postage stamp with > 100 e-"  do you mean on top of the sky, or including that?

**Comment by rmjarvis** [comment]
> But as you build up the flat the BF effect is enough to distort the poisson statistics in this case since some pixels will be higher than others due to fluctuations. 

I don't think so.  At least not measurably.  That's why we need the high flux flats to measure the B/F coefficients.  Just accumulating flats from the sky images doesn't give us enough statistics to see the effect.  You need levels of around 80,000 electrons per pixel to see the correlations.  And even then you need more than one and average over the whole CCD.  

The first order effect of a uniform sky level for B/F is actually nothing.  The pushes and pulls from each pixel all cancel out.  You only get something measurable at second order, from the Poisson variation in the sky level.  So it really is the sigma that matters, and a level of ~800 corresponds to peaks and valleys of ~60 (at 2 sigma).  So that's the relevant number to see how it contributes to the B/F distortions.

So I think putting the threshold at ~100 e- without including the sky is still conservative.  We add on another < 100 e- from the sky variations, which is still well below the point where B/F will begin to show any measurable effect.

**Comment by cwwalter** [comment]
> I don't think so. At least not measurably. That's why we need the high flux flats to measure the B/F coefficients. Just accumulating flats from the sky images doesn't give us enough statistics to see the effect. You need levels of around 80,000 electrons per pixel to see the correlations. And even then you need more than one and average over the whole CCD.

Yes, this is what I mean.  I was trying to say that even at 80,000 you are only looking at a std dev of 280.  It is true that you need to the whole CCD (or many flats) to be able to measure it well over the noise though. 

> 
> The first order effect of a uniform sky level for B/F is actually nothing. The pushes and pulls from each pixel all cancel out. You only get something measurable at second order, from the Poisson variation in the sky level. So it really is the sigma that matters, and a level of ~800 corresponds to peaks and valleys of ~60 (at 2 sigma). So that's the relevant number to see how it contributes to the B/F distortions.

OK.  I was arguing that 100 was still too close to that.  But I agree we can at least think about a threshold and it might be OK.  To really do this iI think we would want a test that we could compare against the full treatment.

> So I think putting the threshold at ~100 e- without including the sky is still conservative. We add on another < 100 e- from the sky variations, which is still well below the point where B/F will begin to show any measurable effect.

So, we can only look at the pixels with the sky (since we draw that first).  So, when you say: "AND there is not yet any pixel in its postage stamp with > 100 e-" do you mean added in quadrature with ~800?

**Comment by rmjarvis** [comment]
> So, we can only look at the pixels with the sky (since we draw that first).

If we are doing the sky first, then I'd suggest the metric `np.any(im.array > np.median(im.array) + 200)`.

200 rather than 100, since something up to around 100 is from the sky itself, so another 100 from sources.

**Comment by rmjarvis** [comment]
cf. https://github.com/lsst/sims_GalSimInterface/pull/70

**Comment by jchiang87** [comment]
I did some runs to test https://github.com/lsst/sims_GalSimInterface/pull/70.  I've attached the `cProfile` outputs for each below:
1. Cori-Haswell, 2 hour run, with Mike's changes.
[haswell_cprofile_mj_faint_object_pr.txt](https://github.com/LSSTDESC/imSim/files/2488070/haswell_cprofile_mj_faint_object_pr.txt)
2. Cori-Haswell, 2 hour run, without Mike's changes.
[haswell_cprofile_baseline_2.txt](https://github.com/LSSTDESC/imSim/files/2488072/haswell_cprofile_baseline_2.txt)
3. Cori-KNL, 10 hour run, with Mike's changes.
[knl_cprofile_10hour.txt](https://github.com/LSSTDESC/imSim/files/2488074/knl_cprofile_10hour.txt)

I used the default settings for `max_flux_simple` and `sensor_limit`.

Interestingly, for the 2 hour Cori-Haswell runs, the version with the changes rendered *fewer* objects, but in other 2 hour runs (same config and inputs), I see performance variations at the 10-30% level.  I've got some lower level tests that isolate timing of the object rendering for the faint guys from things like reading in the SEDs, etc. that I will post later.

Here are some plots comparing runs 1 and 2:
![mj_simple_faint_baseline_comparison](https://user-images.githubusercontent.com/1994473/47097328-0ae0fb80-d1e6-11e8-839b-c6daef126d18.png)
and comparing runs 1 and 3:
![knl_vs_haswell_mj_simple_faint](https://user-images.githubusercontent.com/1994473/47097362-1b917180-d1e6-11e8-8910-5a50b8fa4ef3.png)
I think these are self-explanatory.

**Comment by rmandelb** [comment]
Sorry, but just to be clear: "object rendering time" =  all steps associated with an object, including any initial setups and the actual image rendering?   And the 1 vs. 2 plots show that Mike's branch is worse than master on Haswell when including all of those steps for a given object?

**Comment by jchiang87** [comment]
> "object rendering time" = all steps associated with an object, including any initial setups and the actual image rendering?  And the 1 vs. 2 plots show that Mike's branch is worse than master on Haswell when including all of those steps for a given object?

There are time stamps written to the debug output for each object just before the `drawObject(...)` call in the loop over objects [here](https://github.com/LSSTDESC/imSim/blob/dc2_run2.0_rc/python/desc/imsim/ImageSimulator.py#L360).  The object rendering time is the difference between successive calls given those time stamps.   There are a bunch of per-object steps that take place before that loop, but that call is essentially the start of the rendering (including calculations associated with whether and how an object will be rendered).

**Comment by rmjarvis** [comment]
I think I see what's going on for 1 vs 2.  The changes did reduce the time of the accumulate function by around 25%.  So that's helpful.  But the SED changes were actually detrimental, rather than helpful.  Turns out when the SED isn't a LookupTable, GalSim does some numerical integrals in a few places, so those are now taking up a lot of time.  

Let me try to reproduce this with a toy example on my laptop and see if I can figure out a change that actually ends up speeding things up for objects with flux ~ 10 photons.

**Comment by rmjarvis** [comment]
I just pushed a commit on [PR #70](https://github.com/lsst/sims_GalSimInterface/pull/70) that I think should make a bit difference.  Sorry for not realizing that the solution I put forth would have this big overhead.  

For faint things, switching from a lambda function to a trivial lookup table sped up the faint things by a factor of 100 if they don't use the silicon sensor.  And when they do use the silicon sensor, >99% of the time is in the `accumulate` function now.

**Comment by rmandelb** [comment]
Just to make sure I understand, the SED simplification will affect the DCR in all faint objects, and will affect the e- conversion depth in the faint objects that land on pixels for which the SiliconSensor has to be used?

I should know this off the top of my head... but your default threshold of 100 for the SED simplifications corresponds to typically what magnitude (roughly) in say r or i band?  I'm trying to understand for what fraction of the sample this kicks in. And do I understand correctly that whatever magnorm we've told it to achieve in the band we are simulating will still be met, just with the wrong shape of the SED in that band due to this approximation?

 I am wondering how we can check that there isn't some unforeseen consequence of this change; should we simulate the same sensor 2x, once with `max_flux_simple` at the default and once with `max_flux_simple` set to zero so that it effectively disables your optimization, and look at difference images?  I'm not expecting these approximations to be a problem per se, just that maybe we would want to tune the `max_flux_simple` or rule out any weird bugs.

**Comment by jchiang87** [comment]
The i-band zero points for single visits are around 31.0, so 100 e- would be ~26.

**Comment by rmjarvis** [comment]
Yes, the flux is still right according to the original SED from the cosmoDC2 catalog.  Just the finer details of DCR, and the distribution of conversion depths will be affected.  (The mean DCR effect and the mean conversion depth will be right, but the distributions of each among the photons will be slightly wrong.)

And you can see the object flux on some of Jim's plots above.  There are lots of things < 100 e-.

I'm not sure about your proposed test, since the random number generators will immediately become out of sync in the two runs.  So the details of photon placements will be completely different.  I think we could only compare them in terms of statistics, which is hard.

**Comment by rmandelb** [comment]
> The i-band zero points for single visits are around 31.0, so 100 e- would be ~26.

Thanks.  I should really remember the zero points!  

So the default corresponds to carrying out this simplification for objects that would be S/N<~10 in a 10-year coadd and of course below the detection threshold in individual visits.  And we're really talking about the large majority of objects in the catalogs getting this optimization, so ~95%-ish (estimated by eye from [this plot](https://user-images.githubusercontent.com/6951595/45257763-17656f00-b371-11e8-9d53-e5d7ccc12b4b.png)) will not have the piecewise continuous CatSim SEDs that are being used to carry out solution 4 to the dust/color problem, but rather will (just like solution 4) be forced to the right magnitude in that band while being an even cruder approximation to the SED.  I mean, it's 95%-ish of the objects in the catalogs, but of course it's only a small fraction of the objects we're using for science, and this is just a statement that the catalog and therefore the sample with an approximated SED is dominated by the faint stuff that is around to produce realistic blending to LSST depths.

It feels like a slight philosophical inconsistency to me that we'd say objects down to i~27.5 need to have knots but it's OK to simplify their SEDs if they are fainter than 26.  For the ones between 26-2.75, we are being careful about morphology but less so about the SED shape within the band.  Is the argument essentially that the knots are a 1st order change in how the object appears in the image and the SED shape within the band is more like 2nd order given that the mean DCR and conversion depths are still right? (and DCR and sensor effects are already 1st order effects themselves, so the details are 2nd order)  So that's why it's OK to say there's a range of brightnesses for which we care about knots but are OK approximating the SED within the band?

> I'm not sure about your proposed test, since the random number generators will immediately become out of sync in the two runs.

Yeah - given the brightnesses of the objects affected, I think that test could only work to identify totally wacky bugs and not legitimate changes in their appearance in individual images.

**Comment by rmandelb** [comment]
I'm sorry, we settled on 27 (not 27.5 as I incorrectly stated before) because @fjaviersanchez thought 27 was safer based on blending studies (c.f. [this comment](https://github.com/LSSTDESC/DC2-production/issues/264#issuecomment-421122742)).  So it's just ~1 mag, not 1.5 mag, of overlap.  We could revisit this but I think I'm OK with my 1st vs. 2nd order argument.

**Comment by jchiang87** [comment]
Here's the cProfile output for Mike's recent updates:
[haswell_cprofile_new_simple_faint.txt](https://github.com/LSSTDESC/imSim/files/2493227/haswell_cprofile_new_simple_faint.txt)
and the set of plots comparing the per object timings against sims_GalSimInterface master:
![simple_faint_new_vs_baseline](https://user-images.githubusercontent.com/1994473/47177225-b6647b80-d2cc-11e8-8745-3afa1eb3b2e0.png)
Looks like a ~30-40% speed up for the faint objects.

**Comment by rmjarvis** [comment]
I think it's a bit more than that.  I'm seeing that you got through 167K objects in 2 hours now, compared to 107K in the baseline.  (These are objects that actually made it to a drawImage call, rather than stopping early because 0 photons or off the image.)  So 1.56x as many objects.  Presumably the bright things aren't any faster, so the faint things must be at least 56% faster I would think.

At this point, almost half the time is spent in actual computations that we think we need.  Mostly:
```
335101/168324   13.897    0.000 1947.041    0.012 transform.py:255(shoot)
    63981 1038.351    0.016 1038.351    0.016 sensor.py:247(accumulate)
```
The first is the starting point for the photon shooting, which is mostly the atmospheric PSF shooting.  The second is the accumulation onto the silicon sensor.  You can see that only 64K are using the SiliconSensor, so more than half are being helped by the faint-object branch to use a simple sensor instead.

Here are the next tall poles that look like they could be improved:
```
   172546    8.161    0.000  946.744    0.005 galSimInterpreter.py:639(getHourAngle)
   172546    3.980    0.000  787.180    0.005 core.py:698(sidereal_time)
```
The sidereal time is a constant, so we should be able to do that just once for a whole exposure.  And similarly, for each ccd, the hour angle is basically a constant, so it would not be a big approximation to compute that once per ccd, rather than for each object.

**Comment by rmjarvis** [comment]
It looks like neither of these things use anything from the specific gsObject, so it should be find to just move these lines up to the GalSimInterpreter constructor:
```
        local_hour_angle \
            = self.getHourAngle(self.obs_metadata.mjd.TAI,
                                self.obs_metadata.pointingRA)*galsim.degrees
        obs_latitude = self.observatory.getLatitude().asDegrees()*galsim.degrees
```
and save them as `self.local_hour_angle` and `self.ob_latitude`.

However, I'm not sure how this actually works, since I don't see where `self.observatory` gets set.  So this might be set by imsim after construction?  Since I'm not sure about the data flow here, I'm inclined to just let Jim or Scott make this change.

**Comment by rmjarvis** [comment]
Oh, never mind.  It's a property.  OK.  I made the change on the same branch Jim has been testing.  Should speed things up a bit more.

**Comment by rmjarvis** [comment]
All right, starting to get into the weeds now, but here's another one that looks wasteful:
```
   167464  198.346    0.001  237.839    0.001 galSimDetector.py:30(__init__)
```
Specifically, it seems to be calling the `galsimDetector::__init__` for each object (167K calls).  Can we save on this and just use the same detector object for all objects being drawn on a given detector, rather than make a new one for each object?

I don't see where this is happening though.  It doesn't look like it's in the `drawObject` code (although maybe indirectly somewhere).  Maybe this is in the imsim code?

**Comment by cwwalter** [comment]
It’s here:

https://github.com/lsst/sims_GalSimInterface/blob/master/python/lsst/sims/GalSimInterface/galSimDetector.py#L30

**Comment by cwwalter** [comment]
Or are you asking what calls it?

**Comment by rmjarvis** [comment]
Yeah, why is it getting constructed afresh for each object being drawn?

**Comment by danielsf** [comment]
I suspect it gets called so often because of this implementation of `_newOrigin` in the WCS class

https://github.com/lsst/sims_GalSimInterface/blob/master/python/lsst/sims/GalSimInterface/galSimDetector.py#L151

**Comment by rmjarvis** [comment]
Probably.  Although I couldn't trace a path where that would end up calling GalSimDetector (or make_galsim_detector).  But the `newOrigin` is indeed a similarly tall pole, so it's probably the starting point for this item.
```
   167463    1.380    0.000  242.006    0.001 galSimDetector.py:151(_newOrigin)
```
I switched to a more efficient implementation of this:
```
        _newWcs = GalSim_afw_TanSipWCS.__new__(GalSim_afw_TanSipWCS)
        _newWcs.__dict__.update(self.__dict__)
        _newWcs.crpix1 = origin.x
        _newWcs.crpix2 = origin.y
        _newWcs.fitsHeader.set('CRPIX1', origin.x)
        _newWcs.fitsHeader.set('CRPIX2', origin.y)
        return _newWcs
```

**Comment by rmjarvis** [comment]
I think that might be the end of the easy fixes.  The rest of the big ticket items there look like they need to be called and aren't doing anything too outrageously inefficient.

Of course, there are probably some that are doing some wasteful things, and some might benefit from some C function calls rather than python (e.g. some of the atmospheric stuff is written in pure python, so there might be some things that could be pushed to C++).  But I think any further improvements will probably require some significant development effort in either GalSim or simsCoordUtils to make any headway.

**Comment by jchiang87** [comment]
Here's the cprofile output for the last round of changes:
[haswell_cprofile_00027ed21.txt](https://github.com/LSSTDESC/imSim/files/2494514/haswell_cprofile_00027ed21.txt)
and the plots comparing to master of sims_GalSimInterface:
![simple_faint_00027ed2_comparison](https://user-images.githubusercontent.com/1994473/47198026-cc029100-d31e-11e8-90eb-d30aa8c1259f.png)
Looks like a factor of 2 speedup in the end.   I know these changes are wanted for timing tests at ANL, so I'll proceed with the review.

**Comment by cwwalter** [comment]
For this one, maybe I am mis-understanding the graphs, but it looks like there is a population of objects that take longer to render with the new faint version.  Is that understood? Or am I not understanding the plots?

**Comment by rmjarvis** [comment]
I think there is just some randomness in how long a given object takes.  Probably mostly from the checkpointing I suspect, which involves disk i/o.  

Nothing here should have changed the things with flux > 100 at all, but you can see there is a range of ratios for these objects that scatter around 1.  The mean is probably pretty close to 1, but there is a significant scatter, which just tells you the vicissitudes of this kind of crude timing measure.

For things with flux < 100, there is also a similar scatter, but the mean is significantly < 1.

**Comment by cwwalter** [comment]
@rmjarvis @adrianpope @jchiang87 should we continue to track work here or start a new issue?

**Comment by rmjarvis** [comment]
Either way.  I think here is fine, but we've pretty much addressed all the points mentioned above, so if you'd rather close this and start a fresh issue for further improvements, that's fine too.

**Comment by jchiang87** [comment]
I think we should just close this and open new issues for any further changes we want to make.

**Comment by cwwalter** [comment]
OK I will open a new issue thanks.
